### DISCUSSION TITLE
multiple layers in the Codespaces images

### DISCUSSION THREAD

**Discussion by bartr** [cse-labs/kubernetes-in-codespaces#17] [Q&A] [answered]
Wow that's very interesting regarding layers... I love real world examples like this!

too many time you read blogs, etc saying you 'must always' minimize the number of layers

**Comment by bartr** [comment, marked as answer]
The normal best practice is to minimize the number of layers in a Docker image.

I would suggest that a Codespaces image isn't a "normal" use case. By intentionally creating small layers, we reduced the create time by up to 50% due to the parallel pull.

**Comment by larryclaman** [comment]
I realize this may be straying off topic, but anyway: 

I was the person who wrote the quote in the OP.  I've accepted as 'gospel' for years that you need to minimize the number of docker image layers, but your demo yesterday made me question this.

Yes, I totally get using multistage builds to minimize image size.  But within a single image, what's the actual benefit of minimizing the number of layers?  eg, everyone says to do (A):
```
RUN curl -LO https://releases.hashicorp.com/terraform/${TF_VERSION}/terraform_${TF_VERSION}_linux_amd64.zip \
  && unzip terraform_${TF_VERSION}_linux_amd64.zip \
  && chmod +x ./terraform
```
rather than (B):
```
RUN curl -LO https://releases.hashicorp.com/terraform/${TF_VERSION}/terraform_${TF_VERSION}_linux_amd64.zip
RUN unzip terraform_${TF_VERSION}_linux_amd64.zip
RUN chmod +x ./terraform
```
but there's little discussion as to WHY.  Even Docker [handwaves](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers) on the topic. 

Theres a reasonable Q&A on this topic I found on [Stack Overflow](https://stackoverflow.com/questions/47079114/should-i-minimize-the-number-of-docker-layers)
